### DISCUSSION TITLE
Binding value from child component to parent

### DISCUSSION THREAD

**Discussion by targetlucked69** [solidjs/solid#851] [Q&A]
Hi, in svelte we can bind a child value to a parent like this:

Child component

```html
<script>
    import { onMount } from 'svelte'

    export let username = ''

    onMount(() => {
        username = someAction()
    })
</script>
```

Parent component

```html
<script>
    export let something = ''
</script>

<Child bind:username={something} />
```

Every time username updates, the `something` from parent component will also have its value.

I understand that in order to do this in solid we set the state in parent and pass it together with its' setter in the child component. 

```tsx
const Parent = () => {
    const [something, setSomething] = createSignal('')

    return <Child username={something} setUsername={setSomething} />
}
```

Just want to know if there's another way of doing this...

Another example is this for Vue

```html
<script>
const child = ref()

onMounted(() => {
    console.log(child.value.username) 
})
</script>

<template>
    <Child ref="child" />
</template>
```

**Comment by untlsn-old** [comment]
In Vue child.value is getter and setter 
In Solid getter and setter are separate functions like React useState
But being able to get the setter from getter, or vice versa, would be an interesting option

You can try this way. This is just my idea but I haven't encountered any bugs:

```ts
interface SignalValue<T> {
  (): T
  value: T
}

const createSignalValue = <T>(initValue: T) => {
  const [signal, setSignal] = createSignal(initValue);
  
  const self = () => signal();

  return Object.defineProperty(self, 'value', {
    get: signal,
    set: setSignal
  }) as SignalValue<T>
}
```

Working preview:
https://playground.solidjs.com/?version=1.3.3#NobwRAdghgtgpmAXGGUCWEwBowBcCeADgsrgM4Ae2YZA9gK4BOAxiWGjIbY7gAQi9GcCABM4jXgF9eAM0a0YvADo1aAGzQiAtACsyAegDucAEYqA3EogcuPXgGEAFmjUihEWfMUByAHT7mZ1d3bysbbj4BZiEoXDgAZTQAc2g1KU8FZVUNbT0LKys4Cls+DDjGGShWXkSUqDUANXr6OAAeABVeIrjRMl4oCHwAPn4rXl4ACgBKRF52sd4AN2a4WfmISQKIIpLeZloIMj5ouFiE5NSmtRbeAF5eDq6KHpE+geGJjDRcK5a1qbuIxAC32hz4wDIF3qWF4ZDguFqqQAunc9jE4oj6p9rD8VlNLB5lITQUdYXA1DJUdNAbCoWppgSFkJcEwPAB5Ew6ODMXC+MQyDBwAAK8mIPHwEzhFJh3mW1zg3hhwMJ4yS8NmkLqaiwC3GcNwGvhmLUC0kAKgfWNvza7SGVk2ECsMnoEB5aAODgYEHK1OV4zU8L2Xr49xOZytKwmAAZ8VtxszWZNdQ8nC43MIgy7cLcQPss9J9HbCbGNlt3GJGBNqbcRq17MHxLxCzCRLRmPR4N7fGrcABRAOd3AAIXwAEkRBMVFBCIQVFMS2BJFhwNB4EgwIE0+5qARiOvyFQcHQmKx1+E7AII-L0nJMn59KgMN5GdZveJKtVU8FhCLaIQ+n64yZt6sxXi0rQQB2JjiEWDpOi6boel+6YQBMhCimQszIe4v7-gCgEJowHiTiqDwmPQuC4B6u5wLcKjkZRBwqLwBz2BozAANY5tWIzoX+ZC+HmXZyi0ADUomSEWQFASAfH-oJwYOtJrT6AxVEQFJvAlnB2zFBEvD8lA9BqHw2HCOYi5IkAA

**Comment by thetarnav** [comment]
`atom` helper from solid-use might also be helpful: [solid-use#atom](https://github.com/LXSMNSYC/solid-use#atom)

**Comment by lxsmnsyc** [comment]
Here's a cheap trick

```js
<Child 
  {...{
    get username() {
      return something();
    },
    set username(value) {
      setSomething(value);
    },
  }}
/>
```

then in `Child` you'll do:

```js
props.username = 'John Doe';
```

**Reply by mokeyish** [reply]
> Here's a cheap trick
> 
> ```js
> <Child 
>   {...{
>     get username() {
>       return something();
>     },
>     set username(value) {
>       setSomething(value);
>     },
>   }}
> />
> ```

It' seems not work in `<For>` component.

![图片](https://user-images.githubusercontent.com/16131917/184523650-ad92bbf6-b502-456d-97da-cb280f1af9b2.png)

**Comment by funny-family** [comment]
@walmartwarlord 
Something like this might work for you.
https://playground.solidjs.com/anonymous/73898c4e-5407-446a-a105-875ca478c9a5

**Comment by alexamy** [comment]
Provide a full signal tuple as a prop.

```typescript
import { createSignal, Signal } from 'solid-js';

function Parent() {
  const [count, setCount] = createSignal(0);
  return <Child count={[count, setCount]} />;
}

function Child(props: { count: Signal<number> }) {
  const [count, setCount] = props.count;
  return <></>;
}
```
